src/Kestrel.Core/Internal/Http2/Http2Stream.cs

```diff
@@ -1,1025 +1,54 @@
-
-            if (_onStarting != null)
-            {
-                return InitializeResponseAwaited(firstWriteByteCount);
-            }
-
-            if (_applicationException != null)
-            {
-                ThrowResponseAbortedException();
-            }
-
-            VerifyAndUpdateWrite(firstWriteByteCount);
-
-            return ProduceStart(appCompleted: false);
-        }
-
-        private async Task InitializeResponseAwaited(int firstWriteByteCount)
-        {
-            await FireOnStarting();
-
-            if (_applicationException != null)
-            {
-                ThrowResponseAbortedException();
-            }
-
-            VerifyAndUpdateWrite(firstWriteByteCount);
-
-            await ProduceStart(appCompleted: false);
-        }
-
-        private Task ProduceStart(bool appCompleted)
-        {
-            if (HasResponseStarted)
-            {
-                return Task.CompletedTask;
-            }
-
-            _requestProcessingStatus = RequestProcessingStatus.ResponseStarted;
-
-            return CreateResponseHeader(appCompleted);
-        }
-
-        protected Task TryProduceInvalidRequestResponse()
-        {
-            if (_requestRejectedException != null)
-            {
-                return ProduceEnd();
-            }
-
-            return Task.CompletedTask;
-        }
-
-        protected Task ProduceEnd()
-        {
-            if (_requestRejectedException != null || _applicationException != null)
-            {
-                if (HasResponseStarted)
-                {
-                    // We can no longer change the response, so we simply close the connection.
-                    return Task.CompletedTask;
-                }
-
-                // If the request was rejected, the error state has already been set by SetBadRequestState and
-                // that should take precedence.
-                if (_requestRejectedException != null)
-                {
-                    SetErrorResponseException(_requestRejectedException);
-                }
-                else
-                {
-                    // 500 Internal Server Error
-                    SetErrorResponseHeaders(statusCode: StatusCodes.Status500InternalServerError);
-                }
-            }
-
-            if (!HasResponseStarted)
-            {
-                return ProduceEndAwaited();
-            }
-
-            return WriteSuffix();
-        }
-
-        private async Task ProduceEndAwaited()
-        {
-            await ProduceStart(appCompleted: true);
-
-            // Force flush
-            await Output.FlushAsync(default(CancellationToken));
-
-            await WriteSuffix();
-        }
-
-        private Task WriteSuffix()
+        protected override Task WriteSuffix()
         {
             if (HttpMethods.IsHead(Method) && _responseBytesWritten > 0)
             {
                 Log.ConnectionHeadResponseBodyWrite(ConnectionId, _responseBytesWritten);
             }
```

Comment:
Couldn't the above go in a virtual base method? It looks exactly the same as the log and condition used in Http1Frame.WriteSuffix().

I'm not morally opposed to calling base methods. I get the appeal of the "OnEvent" abstract methods, but I'd rather have calls to base methods than duplicate code like this.

Reply:
I put it in `Frame` (no need to override): 7f15df1ac614401d172569feab45ba842a474edc

Still dupes the HEAD check but that follows the pattern of other methods which have an `Awaited` version. I could eliminate the duplication by moving the check to the top of `WriteSuffix`. It's not a big deal to change the order of log messages in that situation.